### PR TITLE
Adjust Order model to keep info about used voucher code

### PR DESCRIPTION
**Pull request by michal-macioszczyk**
I want to merge this change because Order model needs to keep information about the used voucher code to correctly decrease used value when expire_orders_task is run.
https://github.com/saleor/saleor/issues/13887 

# Impact

- [ ] New migrations
- [ ] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

- [ ] Link to documentation:

# Pull Request Checklist

- [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
- [ ] Database queries are optimized and the number of queries is constant
- [ ] Database migrations are either absent or optimized for zero downtime
- [ ] The changes are covered by test cases

### CODE REVIEW

File: saleor/order/migrations/0175_order_voucher_code.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    dependencies = [
+        ("discount", "0058_voucher_single_use"),
+        ("order", "0174_order_idx_order_created_at"),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name="order",
+            name="voucher_code",
+            field=models.ForeignKey(
```
Comment (IKarbowiak) [inline saleor/order/migrations/0175_order_voucher_code.py:17]: You need to create index for this foreign key directly to use concurrently option

File: saleor/graphql/order/mutations/draft_order_create.py [hunk 2]
```diff
@@ -101,7 +102,13 @@ class DraftOrderInput(BaseInputObjectType):
...
     )
     voucher = graphene.ID(
-        description="ID of the voucher associated with the order.", name="voucher"
+        description="ID of the voucher associated with the order.",
+        name="voucher",
+        deprecation_reason=f"{DEPRECATED_IN_3X_FIELD} Use `voucherCode` instead.",
+    )
+    voucher_code = graphene.ID(
+        description="ID of the voucher code associated with the order.",
+        name="voucherCode",
```
Comment (IKarbowiak) [inline saleor/graphql/order/mutations/draft_order_create.py:111]: In the mutation we need the validation to raise an error when both `voucher` and `voucherCode` are provided

File: saleor/order/tasks.py [hunk 3]
```diff
@@ -45,28 +45,24 @@ def send_order_updated(order_ids):
...
-    # ).values("count")
-
-    # Voucher.objects.filter(
-    #     Exists(voucher_orders),
-    #     usage_limit__isnull=False,
-    # ).annotate(
-    #     order_count=Subquery(count_orders)
-    # ).update(used=F("used") - F("order_count"))
+    voucher_orders = Order.objects.filter(
+        voucher_code__voucher=OuterRef("voucher_id"),
```
Comment (IKarbowiak) [inline saleor/order/tasks.py:49]: `voucher_code__voucher` should be replace with `Exsists` as it's doing `join` on the tables

File: saleor/order/models.py [hunk 4]
```diff
@@ -306,6 +306,10 @@ class Order(ModelWithMetadata, ModelWithExternalReference):
     voucher = models.ForeignKey(
```
Comment (IKarbowiak) [inline saleor/order/models.py:306]: Also we need a discussion here if we want to keep the relation to `voucher` or not

File: saleor/order/migrations/0175_order_voucher_code.py (hunk 1 above)
Comment (michal-macioszczyk) [reply saleor/order/migrations/0175_order_voucher_code.py:17]: done

File: saleor/order/tasks.py (hunk 3 above)
Comment (michal-macioszczyk) [reply saleor/order/tasks.py]: fixed

File: saleor/graphql/order/mutations/draft_order_create.py (hunk 2 above)
Comment (michal-macioszczyk) [reply saleor/graphql/order/mutations/draft_order_create.py:111]: I have added validation for this case

File: saleor/order/migrations/tasks/voucher_to_voucher_code_in_order.py
Comment (IKarbowiak) [inline saleor/order/migrations/tasks/voucher_to_voucher_code_in_order.py:1]: The `tasks` directory is dedicated only to functions that are celery taska, so it shouldn't be here.
In all other mutations we are keeping such methods in the migration file probably to reduce number of files. If you have different idea about that I think we should discuss it with the team first :)

Comment (michal-macioszczyk) [reply saleor/order/migrations/tasks/voucher_to_voucher_code_in_order.py:1]: all fixed and ready for re-review :)

**Review by IKarbowiak — APPROVED** [review]

**Review by SzymJ — APPROVED** [review]

File: saleor/graphql/order/mutations/draft_order_create.py [hunk 5]
```diff
@@ -101,7 +102,13 @@ class DraftOrderInput(BaseInputObjectType):
         description="ID of a selected shipping method.", name="shippingMethod"
     )
     voucher = graphene.ID(
-        description="ID of the voucher associated with the order.", name="voucher"
+        description="ID of the voucher associated with the order.",
+        name="voucher",
+        deprecation_reason=f"{DEPRECATED_IN_3X_FIELD} Use `voucherCode` instead.",
+    )
+    voucher_code = graphene.ID(
+        description="ID of the voucher code associated with the order.",
```
Comment (IKarbowiak) [inline saleor/graphql/order/mutations/draft_order_create.py:110]: Missing `ADDED_IN_X` label

File: saleor/graphql/order/types.py [hunk 6]
```diff
@@ -1178,6 +1183,7 @@ class Order(ModelObjectType[models.Order]):
         deprecation_reason=(f"{DEPRECATED_IN_3X_FIELD} Use `id` instead."),
     )
     voucher = graphene.Field(Voucher)
+    voucher_code = graphene.Field(VoucherCode)
```
Comment (IKarbowiak) [inline saleor/graphql/order/types.py:1186]: I'm not sure if this should be `VoucherCode` field. I mean IMO we want to return `voucher_code` even if the voucher code instance does not exist anymore.

@fowczarek WDYT?

Comment (IKarbowiak) [inline saleor/graphql/order/types.py:1186]: Also missing description and `ADDED_IN_X` label

File: saleor/graphql/order/mutations/draft_order_create.py (hunk 5 above)
Comment (michal-macioszczyk) [reply saleor/graphql/order/mutations/draft_order_create.py]: added

File: saleor/graphql/order/types.py (hunk 6 above)
Comment (michal-macioszczyk) [reply saleor/graphql/order/types.py]: now voucher_code is a String

Comment (michal-macioszczyk) [reply saleor/graphql/order/types.py]: added

**Comment by michal-macioszczyk** [comment]
This task  "Add migration to populate voucher_code from existing voucher field in `OrderDiscount' model" will be done in another ticket to free up functionality that depends on these changes

File: saleor/graphql/order/mutations/draft_order_create.py [hunk 7]
```diff
@@ -198,8 +206,12 @@ def clean_input(cls, info: ResolveInfo, instance, data, **kwargs):
         channel = cls.clean_channel_id(info, instance, cleaned_input, channel_id)
 
         voucher = cleaned_input.get("voucher", None)
+        voucher_code = cleaned_input.get("voucher_code", None)
+        cls.clean_voucher_and_voucher_code(voucher, voucher_code)
         if voucher:
             cls.clean_voucher(voucher, channel)
```
Comment (IKarbowiak) [inline saleor/graphql/order/mutations/draft_order_create.py:212]: We should also assign `voucher_code` when the `voucher` is provided. We should take the first `voucher_code` and assign

Comment (michal-macioszczyk) [reply saleor/graphql/order/mutations/draft_order_create.py]: done

File: saleor/discount/models.py [hunk 8]
```diff
@@ -478,6 +481,7 @@ class Meta:
             ),
             # Orders searching index
             GinIndex(fields=["name", "translated_name"]),
+            BTreeIndex(fields=["voucher_code"], name="orderdiscount_voucher_code_idx"),
```
[the remark this answers is not in the document]
Comment (michal-macioszczyk) [reply saleor/discount/models.py]: done

File: saleor/graphql/order/tests/mutations/test_draft_order_create.py [hunk 9]
```diff
@@ -115,6 +115,108 @@
     """
 
 
+DRAFT_ORDER_CREATE_MUTATION_VOUCHER_CODE = """
```
Comment (IKarbowiak) [inline saleor/graphql/order/tests/mutations/test_draft_order_create.py:118]: I think this is not needed. We should use `DRAFT_ORDER_CREATE_MUTATION`

File: saleor/graphql/order/tests/mutations/test_draft_order_update.py
Comment (IKarbowiak) [inline saleor/graphql/order/tests/mutations/test_draft_order_update.py:1]: Could you also add a test for `draftOrderUpdate` with `voucherCode`?

**Review by IKarbowiak — COMMENTED** [review]
Looks good 🚀 Only small comments in tests

File: saleor/graphql/order/tests/mutations/test_draft_order_create.py (hunk 9 above)
Comment (michal-macioszczyk) [reply saleor/graphql/order/tests/mutations/test_draft_order_create.py]: done

File: saleor/graphql/order/tests/mutations/test_draft_order_update.py
Comment (michal-macioszczyk) [reply saleor/graphql/order/tests/mutations/test_draft_order_update.py:1]: done

File: saleor/graphql/order/tests/mutations/test_draft_order_create.py [hunk 10]
```diff
@@ -115,6 +116,13 @@
     """
 
 
+@pytest.mark.parametrize(
+    "query_str, voucher_type",
+    (
+        (DRAFT_ORDER_CREATE_MUTATION, "voucher"),
+        (DRAFT_ORDER_CREATE_MUTATION, "voucherCode"),
+    ),
+)
```
Comment (IKarbowiak) [inline saleor/graphql/order/tests/mutations/test_draft_order_create.py:125]: Those should be two separate tests.

File: saleor/graphql/order/tests/mutations/test_draft_order_update.py [hunk 11]
```diff
@@ -104,26 +109,40 @@ def test_draft_order_update_voucher_not_available(
     assert error["field"] == "voucher"
 
 
+@pytest.mark.parametrize(
+    "query_str, voucher_type",
+    (
+        (DRAFT_ORDER_UPDATE_MUTATION, "voucher"),
+        (DRAFT_ORDER_UPDATE_MUTATION, "voucherCode"),
+    ),
+)
```
Comment (IKarbowiak) [inline saleor/graphql/order/tests/mutations/test_draft_order_update.py:118]: Same here

**Review by IKarbowiak — COMMENTED** [review]
We are also missing test when both `voucher` and `voucher_code` is provided for `DraftOrderCreate` and `DraftOrderUpdate` mutation

File: saleor/graphql/order/tests/mutations/test_draft_order_update.py (hunk 11 above)
Comment (michal-macioszczyk) [reply saleor/graphql/order/tests/mutations/test_draft_order_update.py]: done

**Review by IKarbowiak — APPROVED** [review]

**Review by SzymJ — APPROVED** [review]

**Pull request merged**